graph/src/components/link_resolver.rs

```diff
@@ -53,19 +77,158 @@ impl LinkResolver for ipfs_api::IpfsClient {
             None => Box::new(cat),
         }
     }
+
+    fn json_stream(
+        &self,
+        link: &Link,
+    ) -> Box<Future<Item = ValueStream, Error = failure::Error> + Send + 'static> {
+        // Discard the `/ipfs/` prefix (if present) to get the hash.
+        let path = link.link.trim_start_matches("/ipfs/").to_owned();
+        let mut stream = self.cat(&path).fuse();
+        let mut buf = BytesMut::with_capacity(1024);
+        // Count the number of lines we've already successfully deserialized.
+        // We need that to adjust the line number in error messages from serde_json
+        // to translate from line numbers in the snippet we are deserializing
+        // to the line number in the overall file
+        let mut count = 0;
+
+        let stream: ValueStream =
+            Box::new(poll_fn(move || -> Poll<Option<Value>, failure::Error> {
+                loop {
+                    if let Some(offset) = buf.iter().position(|b| *b == b'\n') {
+                        let line_bytes = buf.split_to(offset + 1);
```

Comment:
I think this will lead to another complication: when we are processing a bunch of bytes we read, we need to make sure that we do not process the last bit of that chunk unless it happens to align with a line boundary; i.e., if we read a chunk `"stuff"\n"more `, we must not call the JSON callback with `"more `. With the `BufRead.lines` method, we have to check every time we get a line if whether it is the last line in `buf`, and whether that ended in a newline or not. Or we'd have to keep an ending incomplete line somewhere else before creating the `BufReader`. Both options seem to have at least as much overhead as what's there now.